### PR TITLE
Add "resource contention" exercise

### PR DESCRIPTION
**Pull request by andrew-a-dev**
First attempt at a "discussion-based" engineering question, modelled on our recent problems with `motif` and TextRazor.

This is the first thing I've written like this, so I have no idea how useful/valuable it is here.

_FULL DISCLOSURE_: I stole the structure, and lots of copy, from @jplhomer 's excellent EM exercise, here: https://github.com/voxmedia/revenue-hiring/pull/12 🙌

### CODE REVIEW

File: 2-resource-contention/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+1. We'd like to hear a human-worded converversation about how you'd solve this problem.
+1. You are not expected to write, share, or read code at any point in this interview.
+1. Though we do expect that your solution would be technically feasible to create.
+1. If you'd like you can share a block or flow diagram, via screensharing, though this is not necessary.
+
+---
+
+## The Problem
+
+* We want to be able to have as many workers as possible (in our reality, they're `resque` workers, but that's an implementation detail), while still conforming to our API connection limit?
```
Comment (jplhomer) [inline 2-resource-contention/README.md]:
```suggestion
* We want to be able to have as many workers as possible (in our reality, they're `resque` workers, but that's an implementation detail), while still conforming to our API connection limit.
```
Removed question mark

File: 2-resource-contention/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+* The "total number of worker processes" will change over time.
+
+* Assume a (successful) API request will return in less than 5 seconds.
+
+* The worker processes share access to a database (`postgres`, in our case), and a key-value store (`redis`), but have no other inter-process communication ability. (As such, when they start, they have no direct awareness of how many other workers there may be, etc.)
+
+* In reality, incoming load to the service varies considerably over time, but for the sake of this exercise, assume there is _always_ load that requires consuming this API.
+
+* While exceeding the connection limit is "bad" (because it causes exceptions to be raised, which causes other problems), **consistently** having "too few" connections is possibly _worse_ because we're paying for access to this API, and not using it to the maximum potential (a.k.a. "leaving moneytacos on the table").
```
Comment (jplhomer) [inline 2-resource-contention/README.md]: Lol 🌮 ❤️

File: 2-resource-contention/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# Resource Contention Exercise
+
+Our latest project involves many independent worker-processes performing many disparate tasks (e.g. fetching URL contents, making requests to external APIs, calculations on data, etc.).
+
+Our deployment platform (Heroku) makes scaling the resources peforming this work incredibly easy, but sadly, we have other choke-points: one of the external APIs being consumed has a hard-limit on the number of concurrent connections we can make.
```
Comment (banderson623) [inline 2-resource-contention/README.md]:
```suggestion
Our deployment platform makes scaling the resources performing this work incredibly easy, but sadly, we have other choke-points: one of the external APIs being consumed has a hard-limit on the number of concurrent connections we can make.
```

Suggest removing vendor specific name.

File: 2-resource-contention/README.md (hunk 1 above)
Comment (banderson623) [reply 2-resource-contention/README.md]: Another suggestion (which I don't know how to add to the pending suggestion), is that `resque` detail is probably not as important as say "we have `redis` available and these are [`Active::Job`s from Rails](https://guides.rubyonrails.org/active_job_basics.html)"

Comment (banderson623) [reply 2-resource-contention/README.md]: oh you have the redis detail below, nevermind this.

**Comment by banderson623** [comment]
@andrew-a-dev I really appreciate this exercise! The reality of this problem feels _fresh._ And the greenfield opportunity is really a fun place to explore.

My only _question_ is: do we feel like this is an [E2 level](https://docs.google.com/document/d/1b82uAxfeNMTVj0p2_SKI7_QHsDcRMuSZQcV6EiUSD3M/edit) exercise? 

- It might be ideal as-is and this comment is ignorable.
- Would a few more _specific examples and outcomes_ make this solidly an E2 exercise?
- This could include specific rates and limits on each of our systems so the candidate could mentally-simulate how loads flow and bottleneck in the environment.

File: 2-resource-contention/README.md (hunk 1 above)
Comment (hrossvt) [reply 2-resource-contention/README.md]: Another option could be to ditch any reference to redis, resque - and see if the candidate comes up with a queue suggestion on their own

**Comment by hrossvt** [comment]
I like the exercise, thanks for contributing it Drew!

My main feedback builds off of Brian's 3rd bullet:

- _This could include specific rates and limits on each of our systems so the candidate could mentally-simulate how loads flow and bottleneck in the environment._

To help the candidate better visualize this problem overall, we could start by giving a high level overview of CTS, Motif.  "simplify it" by stating the workers do say 2 or 3 things...
for example :

1 - A db lookup that grabs lists related to a url.   We could add in details about how many db connections we use (say a low number so they could list this as an area for imrovement), how fast dn connection is, what operation it does (select, insert) - This could also get the candidate to bring up indexing DBs.

2- An API call to grab data about the url (text razor) - and list out it's limitations - call it out as the bottleneck

**Comment by andrew-a-dev** [comment]
> @andrew-a-dev I really appreciate this exercise! The reality of this problem feels _fresh._ And the greenfield opportunity is really a fun place to explore.
> 
> My only _question_ is: do we feel like this is an [E2 level](https://docs.google.com/document/d/1b82uAxfeNMTVj0p2_SKI7_QHsDcRMuSZQcV6EiUSD3M/edit) exercise?
> 
> * It might be ideal as-is and this comment is ignorable.
> * Would a few more _specific examples and outcomes_ make this solidly an E2 exercise?
> * This could include specific rates and limits on each of our systems so the candidate could mentally-simulate how loads flow and bottleneck in the environment.
> * Would providing a few more specifics/direction ensure this exercise is scoped to no-more-than 60 minutes, even if you haven't thought about this problem before?

**Comment by andrew-a-dev** [comment]
@banderson623 Thanks for this feedback! My questions follow:

> My only _question_ is: do we feel like this is an E2 level exercise?

I'm not sure which way you're suggesting this is "off". (Is it "too much"? "Not enough"?)
 
> * It might be ideal as-is and this comment is ignorable.

Woohoo, `merge`. 😂 

> * Would a few more _specific examples and outcomes_ make this solidly an E2 exercise?

I'm not sure what more examples/outcomes are available to flesh this out in that way. (I think I might just be misunderstanding something here...)

> * This could include specific rates and limits on each of our systems so the candidate could mentally-simulate how loads flow and bottleneck in the environment.

But that sounds like a different problem to solve; this is specifically about the _resource-contention_ issue; the setup **presumes** "_load exceeds instantaneous service capacity_". 

Managing resourcing for a dynamic load is a very very interesting problem, but it's not the one we were solving / I was trying to write-up.

> * Would providing a few more specifics/direction ensure this exercise is scoped to no-more-than 60 minutes, even if you haven't thought about this problem before?

Again, I'm not sure exactly what that would look like in this case. Maybe we should chat about it.

**Comment by andrew-a-dev** [comment]
> * _This could include specific rates and limits on each of our systems so the candidate could mentally-simulate how loads flow and bottleneck in the environment._
>
> [...]
>
> I'm thinking get them to talk about async calls, queues, db optimizations, keeping track of tr concurrent threads, checking for stale connections

Similarly to my comment reply above, I think these are really interesting problems to solve, lots to discuss, and lots of depth to dig into, etc.; but they're not the problems I was trying to write-up for this exercise. So maybe we need to chat about this a bit more.

**Comment by banderson623** [comment]
Thanks for taking the time to respond Drew, and re-reading this, I really enjoy this exercise. My largest concern, poorly stated above, was that this might be too abstract to expect and E2 to reasonable accomplish this in 1 hour.

For full disclosure: our team took much longer (than 1 hour) to determine an ideal system to handle this problem in the real world.

If I take the language from our internal document's definition of E2:
>  Contributed significant work in development at Vox Media and in other jobs, and can take specifications and turn them into functioning software with minimal guidance.
> **Key attributes**: Has shown dependable high quality contribution to technical products and is a good team collaborator. Receives general instructions on routine work and detailed instructions and guidance on new projects or assignments.
> **Business value**: Helps create products, handles development tasks without oversight.

While I expect them to be able to approach straightforward problems, architect, and build solutions, I am not how much they could accomplish in a 1 hour window.


- I wonder if we could limit the scope by providing either more guidance toward a solution
- Limit the scope by adding more specifics of the problem space we are describing
- Possibly, move this out of the abstract and into code, and give them enough guidance and setup-code to be able to construct and stimulate a real system.

**Comment by walia** [comment]
Great work Drew! I like this exercise it provides us a lot of opportunity to discuss big picture problems. 


I was thinking about scoping the problem with more details maybe providing a diagram with specific numbers. The bottleneck in the problem are two external services that can only handle limited number requests than what is coming in. The application has to fetch data from both services for each input and the response JSON from the services is merged into a single JSON and saved into the db. They can design whatever queue system they want. We should still be able to get at the discussion questions.

**Pull request merged**